demos/graph-classification/dgcnn-graph-classification.ipynb

```diff
@@ -0,0 +1,1061 @@
+{
```

Comment:
<p>I think this may not be doing exactly the right thing, because the dimension of the predictions is <code>(batch_size, 13, 1)</code>. For instance, adding a cell before the <code>epochs = 300</code> one that does:</p><p> </p><pre class="ql-syntax" spellcheck="false">preds = model.predict(test_gen)
preds.shape
</pre><p>I see <code>(112, 13, 1)</code> , i.e. there's 13 predictions for each of the 112 test graphs. Is that appropriate for this problem?</p><p> </p><p>https://github.com/stellargraph/stellargraph/issues/1344 is relevant.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/stellargraph/stellargraph/pull/1319/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Reply:
This might be the reason for the flaky performance. 

The issue is the output of the `Conv1D` layer. I tried flattening the output before passing to the `Dense` layer but unfortunately the `tf.keras.layers.Flatten` layer outputs a tensor of size `[None, None]` which cannot be used as input to the `Dense` layer. 

An alternative approach that seems to work is to apply a pooling operation such as `x_out = tf.keras.layers.GlobalMaxPool1D()(x_out)` after the second `Conv1D` layer and before the `Dense` layers. This produces the desired outcome and performance closer to what is reported in the paper; however, I still need to verify the latter. I think that adding `Flatten` or another pooling layer after a `Conv1D` layer and before a `Dense` layer is common practice due to the former's output shape incompatibility with the latter's input.

Reply:
@huonw 

I fixed the above problem by making a small update to the `SortPooling` layer such that when `flatten_output=True` the output tensor is reshaped to have the correct shape instead of `None`. This is possible since the output shape for this layer is known given the inputs. This change makes it possible to follow the `Conv1D` layer with a `Flatten` layer that correctly passes the data to the `Dense` layer avoiding the error I mentioned above.

Performance now is on par with the original paper although I will run a bigger experiment with repeated k-fold cross validation to verify. 

I think the demo is ready for a proper review.

P.